File: R_app_store/datastorekit/R/acceptance_tests/drivers/postgres_driver.R
library(R6)
library(DBI)

PostgresDriver <- R6Class(
  "PostgresDriver",
  inherit = DatabaseDriver,
  public = list(
    orchestrator = NULL,
    datastore_key = NULL,
    conn = NULL,
    key_field = "unique_id",

    initialize = function(orchestrator, datastore_key) {
      self$orchestrator <- orchestrator
      self$datastore_key <- datastore_key
      self$conn <- orchestrator$adapters[[datastore_key]]$connection$conn
    },

    create_table = function(table_name, schema) {
      columns <- sapply(names(schema), function(col) {
        type <- switch(schema[[col]],
                       integer = "INTEGER",
                       character = "VARCHAR",
                       numeric = "FLOAT",
                       stop(sprintf("Unsupported column type: %s", schema[[col]])))
        if (col == self$key_field) {
          sprintf("%s %s PRIMARY KEY", col, type)
        } else {
          sprintf("%s %s", col, type)
        }
      })
      query <- sprintf("CREATE TABLE %s.%s (%s)",
                       self$orchestrator$adapters[[self$datastore_key]]$profile$schema,
                       table_name,
                       paste(columns, collapse = ", "))
      DBI::dbExecute(self$conn, query)
      message(sprintf("Created table %s.%s with PostgreSQL", self$orchestrator$adapters[[self$datastore_key]]$profile$schema, table_name))
    },

    create = function(table_info, data) {
      table <- self$orchestrator$get_table(self$datastore_key, TableInfo$new(
        table_name = table_info$table_name,
        keys = table_info$keys,
        scd_type = table_info$scd_type,
        datastore_key = table_info$datastore_key,
        columns = table_info$columns
      ))
      table$create(data)
    },

    read = function(table_info, filters) {
      table <- self$orchestrator$get_table(self$datastore_key, TableInfo$new(
        table_name = table_info$table_name,
        keys = table_info$keys,
        scd_type = table_info$scd_type,
        datastore_key = table_info$datastore_key,
        columns = table_info$columns
      ))
      table$read(filters)
    },

    update = function(table_info, data, filters) {
      table <- self$orchestrator$get_table(self$datastore_key, TableInfo$new(
        table_name = table_info$table_name,
        keys = table_info$keys,
        scd_type = table_info$scd_type,
        datastore_key = table_info$datastore_key,
        columns = table_info$columns
      ))
      table$update(data, filters)
    },

    delete = function(table_info, filters) {
      table <- self$orchestrator$get_table(self$datastore_key, TableInfo$new(
        table_name = table_info$table_name,
        keys = table_info$keys,
        scd_type = table_info$scd_type,
        datastore_key = table_info$datastore_key,
        columns = table_info$columns
      ))
      table$delete(filters)
    },

    sync_to = function(source_table_info, target_driver, target_table, method) {
      if (method != "full_load") {
        stop(sprintf("Unsupported sync method: %s", method))
      }
      source_table <- source_table_info$table_name
      source_schema <- self$orchestrator$adapters[[self$datastore_key]]$profile$schema
      target_datastore_key <- target_driver$datastore_key
      target_db <- strsplit(target_datastore_key, ":")[[1]][1]
      target_schema <- strsplit(target_datastore_key, ":")[[1]][2]
      self$orchestrator$replicate(
        source_db = self$orchestrator$adapters[[self$datastore_key]]$profile$dbname,
        source_schema = source_schema,
        source_table = source_table,
        target_db = target_db,
        target_schema = target_schema,
        target_table = target_table
      )
    }
  )
)

File: app_store/datastorekit/acceptance_tests/drivers/postgres_driver.py
# datastorekit/acceptance_tests/drivers/postgres_driver.py
from typing import Dict, Any, List
from sqlalchemy import create_engine, Table, MetaData, Column
from ..drivers.database_driver import DatabaseDriver
from datastorekit.orchestrator import DataStoreOrchestrator
from datastorekit.replication.databricks_to_postgres import DatabricksToPostgresReplicator
import logging

logger = logging.getLogger(__name__)

class PostgresDriver(DatabaseDriver):
    def __init__(self, orchestrator: DataStoreOrchestrator, datastore_key: str):
        self.orchestrator = orchestrator
        self.datastore_key = datastore_key
        self.engine = orchestrator.adapters[datastore_key].engine
        self.metadata = MetaData()

    def create_table(self, table_name: str, schema: Dict[str, Any]):
        """Create a table with the given schema."""
        columns = []
        for col_name, col_type in schema.items():
            if col_name == "id":
                columns.append(Column(col_name, col_type, primary_key=True))
            else:
                columns.append(Column(col_name, col_type))
        schema_name = self.orchestrator.adapters[self.datastore_key].profile.schema
        Table(table_name, self.metadata, *columns, schema=schema_name)
        self.metadata.create_all(self.engine)
        logger.info(f"Created table {schema_name}.{table_name} in PostgreSQL")

    def create(self, table_info: Dict[str, str], data: List[Dict[str, Any]]):
        """Insert records into the table."""
        table = self.orchestrator.get_table(self.datastore_key, table_info)
        table.create(data)

    def read(self, table_info: Dict[str, str], filters: Dict[str, Any]) -> List[Dict[str, Any]]:
        """Read records from the table."""
        table = self.orchestrator.get_table(self.datastore_key, table_info)
        return table.read(filters)

    def update(self, table_info: Dict[str, str], data: List[Dict[str, Any]], filters: Dict[str, Any]):
        """Update records in the table."""
        table = self.orchestrator.get_table(self.datastore_key, table_info)
        table.update(data, filters)

    def delete(self, table_info: Dict[str, str], filters: Dict[str, Any]):
        """Delete records from the table."""
        table = self.orchestrator.get_table(self.datastore_key, table_info)
        table.delete(filters)

    def sync_to(self, source_table_info: Dict[str, str], target_driver: 'DatabaseDriver', target_table: str, method: str):
        if method != "full_load":
            raise ValueError(f"Unsupported sync method: {method}")
        source_table = source_table_info["table_name"]
        source_schema = self.orchestrator.adapters[self.datastore_key].profile.schema or "public"
        target_datastore_key = target_driver.datastore_key
        target_db, target_schema = target_datastore_key.split(":")
        self.orchestrator.sync_tables(
            source_db=self.orchestrator.adapters[self.datastore_key].profile.dbname,
            source_schema=source_schema,
            source_table_name=source_table,
            target_db=target_db,
            target_schema=target_schema,
            target_table_name=target_table,
            filters=None,
            chunk_size=1000
        )
        logger.info(f"Synchronized {source_schema}.{source_table} to {target_schema}.{target_table} using {method}")